File: pathfinder-frontend/src/core/Effect.ts
import {MutableDataContext} from "../logic/DataContext";
import {Formula} from "../logic/Formula";
import Resolvable from "../logic/Resolvable";

export default abstract class Effect {

  static setValue(targetKey: string, value: string|number|boolean|Resolvable): Effect {
    return new SetResolvableEffect(targetKey, value);
  }

  static addNumber(targetKey: string, delta: number): Effect {
    return new AddNumberEffect(targetKey, delta);
  }

  static setFormula(targetKey: string, formula: string): Effect {
    return new SetFormulaEffect(targetKey, formula);
  }

  static renameKey(targetKey: string, renamedKey: string): Effect {
    return new RenameKeyEffect(targetKey, renamedKey);
  }

  public abstract applyTo(state: MutableDataContext): void;

  public onlyIf(formula: string|Resolvable): Effect {
    if (formula === '') {
      return this;
    }
    return new ConditionalEffect(this, Formula.parse(formula));
  }
}

export class ConditionalEffect extends Effect {
  constructor(private readonly effect: Effect, private readonly formula: Resolvable) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    if (this.formula.resolve(state)?.asBoolean() ?? false) {
      this.effect.applyTo(state);
    }
  }
}

export class SetNumberEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly value: number) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    state.set(this.targetKey, this.value);
  }
}

export class SetTextEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly value: string) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    state.set(this.targetKey, this.value);
  }
}

export class SetFormulaEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly formula: string) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    state.set(this.targetKey, Formula.parse(this.formula));
  }
}

export class SetResolvableEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly resolvable: string|number|boolean|Resolvable) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    state.set(this.targetKey, this.resolvable);
  }
}

export class AddNumberEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly delta: number) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    let current = state.get(this.targetKey)?.asNumber() ?? 0;
    state.set(this.targetKey, current + this.delta);
  }
}

export class RenameKeyEffect extends Effect {
  constructor(public readonly targetKey: string, public readonly renamedKey: string) {
    super();
  }

  applyTo(state: MutableDataContext): void {
    state.rename(this.targetKey, this.renamedKey);
  }
}

File: pathfinder-generator/src/main/java/pathfinder/model/v4/Effect.java
package pathfinder.model.v4;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import pathfinder.data.v4.EffectDbo;
import pathfinder.data.v4.EffectDbo.AddActionDbo;
import pathfinder.data.v4.EffectDbo.RenameKeyDbo;
import pathfinder.data.v4.EffectDbo.SetActionDbo;
import pathfinder.model.Id;

public abstract class Effect {

    public static Effect setFormula(Id targetKey, String formula) {
        return setFormula(targetKey.string(), formula);
    }

    public static Effect setNumber(Id targetKey, int value) {
        return setNumber(targetKey.string(), value);
    }

    public static Effect addNumber(Id targetKey, int delta) {
        return addNumber(targetKey.string(), delta);
    }

    public static Effect setFormula(String targetKey, String formula) {
        return new SetFormulaEffect(targetKey, "", formula);
    }

    public static Effect setNumber(String targetKey, int value) {
        return new SetNumberEffect(targetKey, "", value);
    }

    public static Effect addNumber(String targetKey, int delta) {
        return new AddNumberEffect(targetKey, "", delta);
    }

    public static Effect renameKey(String targetKey, String renamedKey) {
        return new RenameKeyEffect(targetKey, renamedKey, "");
    }

    public static Effect renameKey(Id targetKey, Id renamedKey) {
        return new RenameKeyEffect(targetKey.string(), renamedKey.string(), "");
    }

    public abstract Effect onlyIf(String condition);
    public abstract EffectDbo toDbo();

    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    private static class SetFormulaEffect extends Effect {
        private final String targetKey;
        private final String condition;
        private final String formula;

        @Override
        public Effect onlyIf(String condition) {
            return new SetFormulaEffect(targetKey, condition, formula);
        }

        @Override
        public EffectDbo toDbo() {
            return EffectDbo.newBuilder()
                    .setTargetKey(targetKey)
                    .setCondition(condition)
                    .setSetAction(SetActionDbo.newBuilder()
                            .setFormula(formula)
                            .build())
                    .build();
        }
    }

    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    private static class SetNumberEffect extends Effect {
        private final String targetKey;
        private final String condition;
        private final int value;

        @Override
        public Effect onlyIf(String condition) {
            return new SetNumberEffect(targetKey, condition, value);
        }

        @Override
        public EffectDbo toDbo() {
            return EffectDbo.newBuilder()
                    .setTargetKey(targetKey)
                    .setCondition(condition)
                    .setSetAction(SetActionDbo.newBuilder()
                            .setNumberValue(value)
                            .build())
                    .build();
        }
    }

    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    private static class AddNumberEffect extends Effect {
        private final String targetKey;
        private final String condition;
        private final int delta;

        @Override
        public Effect onlyIf(String condition) {
            return new AddNumberEffect(targetKey, condition, delta);
        }

        @Override
        public EffectDbo toDbo() {
            return EffectDbo.newBuilder()
                    .setTargetKey(targetKey)
                    .setCondition(condition)
                    .setAddAction(AddActionDbo.newBuilder()
                            .setNumberDelta(delta)
                            .build())
                    .build();
        }
    }

    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    private static class RenameKeyEffect extends Effect {
        private final String targetKey;
        private final String renamedKey;
        private final String condition;

        @Override
        public Effect onlyIf(String condition) {
            return new RenameKeyEffect(targetKey, renamedKey, condition);
        }

        @Override
        public EffectDbo toDbo() {
            return EffectDbo.newBuilder()
                    .setTargetKey(targetKey)
                    .setCondition(condition)
                    .setRenameAction(RenameKeyDbo.newBuilder()
                            .setRenamedKey(renamedKey)
                            .build())
                    .build();
        }
    }
}
